### PR TITLE
Unify error handling in Azure services

### PR DESCRIPTION
**Pull request by Flanker32**
- Unified way of delivering error to user, for [#1791921](https://dev.azure.com/mseng/VSJava/_workitems/edit/1791921)
- Handle operations in closure in exception handler [#1794805](https://dev.azure.com/mseng/VSJava/_workitems/edit/1794805)
- New dialog for foreground exceptions [#1794804](https://dev.azure.com/mseng/VSJava/_workitems/edit/1794804)

**Default**
image

**Details**
image

### CODE REVIEW

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/common/handler/IntelliJAzureExceptionHandler.java [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+        final Notification notification = new Notification(NOTIFICATION_GROUP_ID, AZURE_TOOLKIT_ERROR, body, NotificationType.ERROR);
+        for (AzureExceptionAction exceptionAction : actions) {
+            notification.addAction(new AnAction(exceptionAction.name()) {
+                @Override
+                public void actionPerformed(@NotNull final AnActionEvent anActionEvent) {
+                    exceptionAction.actionPerformed(throwable);
+                }
+            });
+        }
+        AzureTaskManager.getInstance().runLater(() -> Notifications.Bus.notify(notification, project));
```
Comment (wangmingliang-ms) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/common/handler/IntelliJAzureExceptionHandler.java:128]: it's dangerous(maybe infinite recursion) if `notify` throws exceptions here

File: Utils/azuretools-core/src/com/microsoft/azuretools/core/mvp/model/AzureMvpModel.java [hunk 2]
```diff
@@ -84,6 +86,11 @@ public Subscription getSubscriptionById(String sid) {
         return ret;
     }
 
+    @AzureOperation(value = "Test exception with input `%s`", params = {"$input"}, type = AzureOperation.Type.SERVICE)
+    public void testException(String input) {
+        throw new AzureToolkitRuntimeException("Test exception", null , "Please retry", "001");
+    }
+
```
Comment (wangmingliang-ms) [inline Utils/azuretools-core/src/com/microsoft/azuretools/core/mvp/model/AzureMvpModel.java:93]: remove this?

File: Utils/azure-explorer-common/src/com/microsoft/tooling/msservices/serviceexplorer/NodeActionListener.java [hunk 3]
```diff
@@ -79,8 +79,9 @@ protected abstract void actionPerformed(NodeActionEvent e)
             EventUtil.logEvent(EventType.info, operation, buildProp(node));
             actionPerformed(e);
             return Futures.immediateFuture(null);
-        } catch (AzureCmdException ex) {
+        } catch (AzureCmdException | RuntimeException ex) {
             EventUtil.logError(operation, ErrorType.systemError, ex, null, null);
+            AzureExceptionHandler.getInstance().handleException(ex, false);
             return Futures.immediateFailedFuture(ex);
```
Comment (wangmingliang-ms) [inline Utils/azure-explorer-common/src/com/microsoft/tooling/msservices/serviceexplorer/NodeActionListener.java:85]: what if the owner of the future instance handle the exception again.

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/webapp/webappconfig/WebAppRunState.java [hunk 4]
```diff
@@ -74,9 +75,10 @@ public WebAppRunState(Project project, WebAppConfiguration webAppConfiguration)
 
     @Nullable
     @Override
-    @AzureOperation(value = "run web app", type = AzureOperation.Type.ACTION)
+    @AzureOperation(value = "deploy web app with configuration", type = AzureOperation.Type.ACTION)
     public WebAppBase executeSteps(@NotNull RunProcessHandler processHandler
         , @NotNull Map<String, String> telemetryMap) throws Exception {
+        AzureMvpModel.getInstance().testException("wrong input");
```
Comment (wangmingliang-ms) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/webapp/webappconfig/WebAppRunState.java:81]: remove

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/AzureRunProfileState.java [hunk 5]
```diff
@@ -100,6 +102,7 @@ protected Operation createOperation() {
 
     protected void onFail(@NotNull Throwable error, @NotNull RunProcessHandler processHandler) {
         onFail(error.getMessage(), processHandler);
+        IntelliJAzureExceptionHandler.getInstance().handleException(project, new AzureToolkitRuntimeException("execute run configuration", error), true);
```
Comment (wangmingliang-ms) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/AzureRunProfileState.java:105]: seems better to move to the next `onFail`

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/lib/webapp/WebAppService.java [hunk 6]
```diff
@@ -61,6 +62,7 @@ public WebApp createWebApp(final WebAppConfig config) {
         settings.setCreatingNew(true);
         final Map<String, String> properties = settings.getTelemetryProperties(null);
         final Operation operation = TelemetryManager.createOperation(WEBAPP, CREATE_WEBAPP);
+        AzureMvpModel.getInstance().testException("exception input");
```
Comment (wangmingliang-ms) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/lib/webapp/WebAppService.java:65]: remove

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/AzureRunProfileState.java (hunk 5 above)
Comment (Flanker32) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/AzureRunProfileState.java:110]: Refactor done

File: Utils/azure-explorer-common/src/com/microsoft/tooling/msservices/serviceexplorer/NodeActionListener.java (hunk 3 above)
Comment (Flanker32) [reply Utils/azure-explorer-common/src/com/microsoft/tooling/msservices/serviceexplorer/NodeActionListener.java:85]: It was called by [fireNodeActionEvent](https://github.com/microsoft/azure-tools-for-java/blob/develop/Utils/azure-explorer-common/src/com/microsoft/tooling/msservices/serviceexplorer/NodeAction.java#L69), which only handle the after event listener

**Review by wangmingliang-ms — APPROVED** [review]

**Pull request merged**